File: correlation_coefficient.R
correlation_coef <- function(x, y, n) {
# Sort x and y by x
  sorted_indices <- order(x)
  x_sorted <- x[sorted_indices]
  y_sorted <- y[sorted_indices]
  # Remove records whose values of x are outside the 0.01-0.99 quantile range to avoid outliers
  a <- quantile(x, 0.01)
  b <- quantile(x, 0.99)
  # Spliting into intervals by x
  intervals <- seq(a, b, length.out = n + 1)
  # Calculate local variances
  local_variances <- numeric(n)
  for (i in 1:n) {
    x_interval <- x_sorted[x_sorted >= intervals[i] & x_sorted < intervals[i + 1]]
    y_interval <- y_sorted[x_sorted >= intervals[i] & x_sorted < intervals[i + 1]]
    local_variances[i] <- var(y_interval) * length(y_interval)
  }
# Calculate coefficient
  total_variance <- var(y)
  total_points <- length(y)
  dispersion_ratio <- sum(local_variances) / total_points / total_variance
  return(dispersion_ratio)
}

File: correlation_coefficient.py
def corr_coef(df: pd.DataFrame, y: str, x: str, n = 100):
    # Remove records whose values of x are outside the 0.01-0.99 quantile range to avoid outliers
    quantile_min = df[x].quantile(0.001)
    quantile_max = df[x].quantile(0.999)
    df_filtered = df[df[x].between(quantile_min, quantile_max)]
    
    # Calculate step size
    steps_count = int(len(df_filtered) / n) + 1
    step = (df_filtered[x].max() - df_filtered[x].min()) / steps_count
 
    # Calculate local variances
    y_vars = []
    for i in range(steps_count):
        x_min = quantile_min + step * i
        x_max = quantile_min + step * (i + 1)
        y_values = df_filtered[y][(df_filtered[x] >= x_min) & (df_filtered[x] < x_max)]
        y_count = len(y_values)
        y_vars.append(y_values.var() * y_count)
 
    # Calculation general variance and sum of local variances
    df_var = df_filtered[y].var()
    cov = sum(y_vars)
 
    # Return coef value
    return (1 - cov / (len(df_filtered) * df_var))
